version: '3'

x-meltano-image: &meltano-image
  image: meltano/meltano:latest-python3.10
  command: lock --update --all
  volumes:
    - .:/project

services:
  postgres:
    image: postgres:12
    env_file:
      - .env
    environment:
      POSTGRES_DB: elt_meltano
      POSTGRES_USER: ${POSTGRES_USER}
      POSTGRES_PASSWORD: ${POSTGRES_PASSWORD}
    ports:
      - "5432:5432"
    volumes:
      - ./dbdata:/var/lib/postgresql/data
      - ./data/northwind.sql:/docker-entrypoint-initdb.d/northwind.sql
    networks:
      docker_network:
        ipv4_address: 172.19.0.20
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -U ${POSTGRES_USER} -d elt_meltano"]
      interval: 5s
      timeout: 5s
      retries: 5

  meltano-install:
    <<: *meltano-image
    command: install
    networks:
      docker_network:
    depends_on:
      postgres:
        condition: service_healthy

  airflow-scheduler:
    <<: *meltano-image
    command: invoke airflow scheduler
    networks:
      docker_network:
    depends_on:
      postgres:
        condition: service_healthy
      meltano-install:
        condition: service_completed_successfully
    restart: unless-stopped

  airflow-webserver:
    <<: *meltano-image
    command: invoke airflow webserver
    networks:
      docker_network:
    depends_on:
      postgres:
        condition: service_healthy
      meltano-install:
        condition: service_completed_successfully
    ports:
      - "8080:8080"
    restart: unless-stopped


networks:
  docker_network:
      driver: bridge
      ipam:
        config:
          - subnet: 172.19.0.0/16
